tutorials/python/python-type-driven-development/tutorial.md

```diff
@@ -0,0 +1,209 @@
+twitter-profile: po5i
+github-profile: po5i
+description: Learn how to improve the software quality and readability in our Python code by using the built-in typing system and complementing with other software development processes like type-driven development and test-driven development.
+---
+
+_Type-Driven Development_ is a simple approach that states to add static types to our code that are aligned with the logic. It brings many benefits such as increasing the performance, accuracy, testability, readability, and extensibility of your code by taking advantage of the language’s type system. There are many books that you can use to learn how to achieve this, for example, [“Type-Driven Development with Idris”][tdd-idris] and extrapolate these practices to other languages.
+
+In a [previous article][types-vs-tests], we discussed the benefits of working with _Type-Driven Development_ and the importance of static type checking in compilation time for strongly typed languages such as Java or Haskell. It doesn’t matter if the paradigm is object-oriented or functional, type-checking adds a layer of safety, reducing the amount of tests we need to write, and reduces our dependency on tests.
+
+## What about Python?
+
+Is it possible to use a strong type system in a language like Python? The answer is yes. [PEP-484][pep-484] introduced _Type Annotations to Python_ 3.5 back in 2014. It’s been a while, right? Why is this not popular? Because Python doesn’t require the code has types everywhere to run, they’re completely optional and mostly used by editors and IDEs. Another reason is that Python code is well-known to be readable in the least number of possible lines of code, but adding types to your code doesn’t make it more readable, what a paradox!
+
+In this article, we’ll be walking through the Python type system, how to add types to our code and Docstrings and perform static type checking using [mypy][mypy]. There are plenty options for static type-checking in Python, but in this tutorial we will follow [Guido van Rossum suggestion][guido-mypy].
+
+### Mypy
+
+Mypy is a third-party Python library that provides optional static type checking. Unlike other non-dynamic programming languages like Java, where the static type-checking takes place at compilation time, Mypy CLI does the type-check to a file (or a set of files), so it's up to you to decide where in your workflow the type-checking will be executed, it can be on CI or a pre-commit hook, just to mention some.
+
+To start using mypy, install it using any version of `pip` globally or in your virtual environment.
+
+```bash
+pip install mypy
+```
+
+To perform static type checking to a source file:
+
+```bash
+mypy my_file.py
+```
+
+## Primitive types
+
+Python is a multi-paradigm dynamic language, so the type of a variable is determined based on the value. This might be confusing, but it doesn't mean that Python has _weak typing_. Python is _strong typed_. Surprised? Check this out:
```

Comment:
strongly-typed?

Reply:
Yes, this is confusing for everyone at the beginning. 

Python is a dynamic language because you can do this

```
foo = 4
```

and then foo will be an `int` in runtime. But it's also a strong typing language since you can't do this

```
>>> foo + '2'
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for +: 'int' and 'str'
```
The same way you can in JS for example (which is a weak typing language)